src/riak_kv_eleveldb_backend.erl

```diff
@@ -454,47 +453,35 @@ fold_indexes_fun(FoldIndexFun) ->
             end
     end.
 
+build_list({_K, _V}=KV, Acc) ->
+    [KV | Acc].
+
 range_scan(FoldIndexFun, Buffer, Opts, #state{fold_opts=_FoldOpts,
                                               ref=Ref}) ->
     {_, Bucket, Qry} = proplists:lookup(index, Opts),
     ?SQL_SELECT{'WHERE'    = W,
-                helper_mod = Mod,
-                local_key  = LK} = Qry,
-    %% this is all super-fugly
+                helper_mod = _Mod,
+                local_key  =  #key_v1{ast = LKAST},
+                partition_key = #key_v1{ast = _PKAST}} = Qry,
     {startkey, StartK} = proplists:lookup(startkey, W),
     {endkey,   EndK}   = proplists:lookup(endkey, W),
-    {filter,   Filter} = proplists:lookup(filter, W),
-    StartInclusive = case proplists:lookup(start_inclusive, W) of
-                         none  -> [];
-                         STuple -> [STuple]
-                     end,
-    EndInclusive = case proplists:lookup(end_inclusive, W) of
-                       none  -> [];
-                       ETuple -> [ETuple]
-                   end,
-    AdditionalOptions = lists:flatten(StartInclusive ++ EndInclusive),
-    AdditionalOptions2 =
-        case Filter of
-            [] -> AdditionalOptions;
-            _  -> [{range_filter, Filter} | AdditionalOptions]
-        end,
-    StartK2 = [{Field, Val} || {Field, _Type, Val} <- StartK],
-    StartK3 = riak_ql_ddl:make_key(Mod, LK, StartK2),
-    StartK4 = riak_kv_ts_util:encode_typeval_key(StartK3), %% TODO: Avoid adding/removing type info
-    StartKey = to_object_key(Bucket, StartK4),
-    EndK2 = [{Field, Val} || {Field, _Type, Val} <- EndK],
-    EndK3 = riak_ql_ddl:make_key(Mod, LK, EndK2),
-    EndK4 = riak_kv_ts_util:encode_typeval_key(EndK3),
-    EndKey = to_object_key(Bucket, EndK4),
-    FoldFun = fun({K, V}, Acc) ->
-                      [{K, V} | Acc]
-              end,
+    LocalKeyLen = length(LKAST),
+    StartKey1 = key_prefix(Bucket,  [Value || {_Name,_Type,Value} <- StartK], LocalKeyLen),
+    EndKey1 = key_prefix(Bucket,  [Value || {_Name,_Type,Value} <- EndK], LocalKeyLen),
+    case lists:member({start_inclusive, false}, W) of
```

Comment:
Would you mid rewriting this with StartKey2 outside of the case statement to make it cleaner.
Is there a performance reason you went with ```lists:member``` rather than something like proplists:get_value?  It took me a few minutes to work out why the extra bytes were being added to the true case.

A few lines of code are worth 1000 words
```
StartKey2 = case proplists:get_value(start_inclusive, W, true) of
   false -> <<StartKey1/binary, 16#ff:8>>; % Append extra binary for exclusive start after the key
   true -> StartKey1
end
```